rust/module/src/lib.rs

```diff
@@ -0,0 +1,355 @@
+    }
+}
+
+fn expect_literal(it: &mut token_stream::IntoIter) -> String {
+    if let TokenTree::Literal(literal) = it.next().unwrap() {
+        literal.to_string()
+    } else {
+        panic!("Expected Literal");
+    }
+}
+
+fn expect_group(it: &mut token_stream::IntoIter) -> Group {
+    if let TokenTree::Group(group) = it.next().unwrap() {
+        group
+    } else {
+        panic!("Expected Group");
+    }
+}
+
+fn expect_end(it: &mut token_stream::IntoIter) {
+    if let None = it.next() {
+    } else {
+        panic!("Expected end");
+    }
+}
+
+fn get_ident(it: &mut token_stream::IntoIter, expected_name: &str) -> String {
+    assert_eq!(expect_ident(it), expected_name);
+    assert_eq!(expect_punct(it), ':');
+    let ident = expect_ident(it);
+    assert_eq!(expect_punct(it), ',');
+    ident
+}
+
+fn get_literal(it: &mut token_stream::IntoIter, expected_name: &str) -> String {
+    assert_eq!(expect_ident(it), expected_name);
+    assert_eq!(expect_punct(it), ':');
+    let literal = expect_literal(it);
+    assert_eq!(expect_punct(it), ',');
+    literal
+}
+
+fn get_group(it: &mut token_stream::IntoIter, expected_name: &str) -> Group {
+    assert_eq!(expect_ident(it), expected_name);
+    assert_eq!(expect_punct(it), ':');
+    let group = expect_group(it);
+    assert_eq!(expect_punct(it), ',');
+    group
+}
+
+fn get_byte_string(it: &mut token_stream::IntoIter, expected_name: &str) -> String {
+    let byte_string = get_literal(it, expected_name);
+
+    assert!(byte_string.starts_with("b\""));
+    assert!(byte_string.ends_with("\""));
+
+    byte_string[2..byte_string.len() - 1].to_string()
+}
+
+fn build_modinfo_builtin_string(module_name: &str, name: &str, string: &str) -> String {
+    format!(
+        "
+            // Built-in modules prefix their modinfo strings by `module_name.`
+            #[cfg(not(MODULE))]
+            #[link_section = \".modinfo\"]
+            #[used]
+            pub static {name}: [u8; {length}] = *b\"{module_name}.{string}\\0\";
+        ",
+        module_name = module_name,
+        name = name,
+        length = module_name.len() + 1 + string.len() + 1,
+        string = string,
+    )
+}
+
+fn build_modinfo_string(module_name: &str, name: &str, string: &str) -> String {
+    format!(
+        "
+            // Loadable modules' modinfo strings go as-is
+            #[cfg(MODULE)]
+            #[link_section = \".modinfo\"]
+            #[used]
+            pub static {name}: [u8; {length}] = *b\"{string}\\0\";
+        ",
+        name = name,
+        length = string.len() + 1,
+        string = string,
+    ) + &build_modinfo_builtin_string(module_name, name, string)
+}
+
+/// Declares a kernel module.
+///
+/// The `typename` argument should be a type which implements the [`KernelModule`] trait.
+/// Also accepts various forms of kernel metadata.
+///
+/// Example:
+/// ```rust,no_run
+/// use kernel::prelude::*;
+///
+/// module!{
+///     typename: MyKernelModule,
```

Comment:
`typename` doesn't feel like quite the right  name for this.

Reply:
Initially I used `type_` (with the underscore to avoid the keyword), then felt the `_` was a bit off and changed it. I'm not sure what would be best to use. I could remove it entirely like you had in the previous macro, for the type and the name since those are mandatory, but then it felt odd that some parameters didn't have it... I also considered removing all of the `x:` and then having everything mandatory as a positional parameter (since anyway everything should have a license and most likely author and description too), but then too many positional ones look odd too...